contrib/babelfishpg_tsql/src/multidb.c

```diff
@@ -667,6 +676,122 @@ rewrite_relation_walker(Node *node, void *context)
+	{
+		Node* n = linitial(targList);
+		Node* n_from = linitial(fromList);
+		if(IsA(n, ResTarget) && IsA(n_from, RangeSubselect))
+		{
+			ResTarget* rt = (ResTarget*) n;
+			RangeSubselect* rs = (RangeSubselect*) n_from;
+			if(IsA(rt->val, FuncCall) && IsA(rs->subquery, SelectStmt))
+			{
+				FuncCall* json_mod_fc = (FuncCall*) rt->val;
+				SelectStmt* from_sel_stmt = (SelectStmt*) rs->subquery;
+				rewrite_plain_name(json_mod_fc->funcname);
+				if(is_json_modify(json_mod_fc->funcname) && is_select_for_json(from_sel_stmt))
+				{
+					Node *n = lfourth(json_mod_fc->args);
+					A_Const *escape = (A_Const*) n;
+					escape->val.boolval.boolval = true;
+				}
+			}
+		}
+	}
+}
+
+bool
+is_json_modify(List *name)
+{
+    switch(list_length(name))
+    {
+        case 1:
+        {
+            Node *func = (Node *) linitial(name);
+            if(strcmp("json_modify", strVal(func)) == 0)
+                return true;
+            return false;
+        }
+        case 2:
+        {
+            Node *schema = (Node *) linitial(name);
+            Node *func = (Node *) lsecond(name);
+            if(strcmp("sys", strVal(schema)) == 0 &&
+                strcmp("json_modify", strVal(func)) == 0)
+                return true;
+            return false;
+        }
+        default:
+            return false;
+    }
+}
+
+/*
+ * is_select_for_json takes in a select statement
+ * returns true if the target function call is for json and array_wrapper is false
+ * returns false otherwise
+ */
+bool is_select_for_json(SelectStmt *stmt)
+{
+    List* targetList = stmt->targetList;
+    ListCell* lc = (ListCell*) targetList->elements;
+    Node* n = lfirst(lc);
+    if(IsA(n, ResTarget))
+    {
+        ResTarget* rt = (ResTarget*) n;
+        if(IsA(rt->val, FuncCall))
+        {
+            FuncCall* fc = (FuncCall*) rt->val;
+            return is_for_json(fc);
+        }
+    }
+    return false;
+}
+
+bool
+is_for_json(FuncCall *fc)
+{
+    // In this case, we need to check that the function name is correct, and also that the without_array_wrapper param is not true
+    List* funcname = fc->funcname;
+    List* fc_args = fc->args;
+
+    switch(list_length(funcname))
+    {
+        case 2:
+        {
+            Node *schema = (Node *) linitial(funcname);
+            Node *func = (Node *) lsecond(funcname);
+            if(strcmp("sys", strVal(schema)) == 0 &&
+                strcmp("tsql_select_for_json_result", strVal(func)) == 0)
+            {
+                // If without array wrapper is true, we want to keep the escape characters so we return false
+                return !get_array_wrapper(fc_args);
+            }
+            return false;
+        }
+        default:
+            return false;
+    }
+}
+
+bool get_array_wrapper(List* for_json_args) {
+    FuncCall* agg_fc = (FuncCall *) linitial(for_json_args);
+    List* agg_fc_args = agg_fc->args;
+
+    Node* arr_wrap = lfourth(agg_fc_args);
```

Comment:
Can this function be called in any case where `lfourth(agg_fc_args)` does not exist?

Reply:
The parser rule is set up so anytime JSON_MODIFY is called it calls the TsqlJsonModifyMakeFuncCall function which adds assembles the 4 element arg list so it will always have 4 arguments

Reply:
This function is not defined as static, so it could possibly be referenced in other parts of the code. I would add a check on the list length, and possibly also more checks so that even if this is passed an invalid List, it will error rather than crash.